### PR TITLE
impl(GCS+gRPC): when possible, enable gRPC metrics

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #13998 

This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14265)

### CODE REVIEW

File: google/cloud/storage/google_cloud_cpp_storage_grpc.cmake [hunk 1]
```diff
@@ -233,11 +233,20 @@ if (GOOGLE_CLOUD_CPP_ENABLE_CTYPE_CORD_WORKAROUND)
...
+    target_compile_definitions(
+        google_cloud_cpp_storage_grpc
+        PRIVATE GOOGLE_CLOUD_CPP_STORAGE_AUTO_OTEL_METRICS)
+    target_link_libraries(
+        google_cloud_cpp_storage_grpc
+        PUBLIC google-cloud-cpp::opentelemetry gRPC::grpcpp_otel_plugin
+               opentelemetry-cpp::metrics)
+    # TODO(#14260) - remove this workaround. gRPC does not publish a pkgconfig
+    # module for grpcpp_otel_plugin.
+    set(EXTRA_MODULES "google_cloud_cpp_opentelemetry" "opentelemetry_metrics")
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/storage/google_cloud_cpp_storage_grpc.cmake:248]: https://github.com/googleapis/google-cloud-cpp/pull/14266

File: google/cloud/storage/internal/grpc/default_options.cc [hunk 2]
```diff
@@ -53,6 +54,20 @@ int DefaultGrpcNumChannels(std::string const& endpoint) {
...
 
+bool GrpcEnableMetricsIsSafe() {
+#ifndef GRPC_CPP_VERSION_MAJOR
+  return false;
+#else
+  return (
+      GRPC_CPP_VERSION_MAJOR >= 1 &&
+      (                                                                    //
+          (GRPC_CPP_VERSION_MINOR == 63 && GRPC_CPP_VERSION_PATCH > 1) ||  //
+          (GRPC_CPP_VERSION_MINOR == 64 && GRPC_CPP_VERSION_PATCH > 1) ||  //
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/default_options.cc:65]: I am confused here. I only see tags for `1.63.0`, `1.64.0`.  https://github.com/grpc/grpc/tags

We are assuming that they will add a `1.63.1` that won't fix the problem, then add a `1.63.2` that will fix the problem?

File: google/cloud/storage/internal/grpc/default_options_test.cc [hunk 3]
```diff
@@ -120,12 +121,22 @@ TEST(DefaultOptionsGrpc, DefaultOptionsUploadBuffer) {
...
                         absl::nullopt);
   auto const options = DefaultOptionsGrpc(Options{});
-  EXPECT_TRUE(options.get<storage_experimental::EnableGrpcMetricsOption>());
+  auto const expected =
+      (GRPC_CPP_VERSION_MAJOR >= 1 &&
+       (                                                                    //
+           (GRPC_CPP_VERSION_MINOR == 63 && GRPC_CPP_VERSION_PATCH > 1) ||  //
+           (GRPC_CPP_VERSION_MINOR == 64 && GRPC_CPP_VERSION_PATCH > 1) ||  //
+           (GRPC_CPP_VERSION_MINOR >= 65)                                   //
+           ));
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/default_options_test.cc:135]: I suggest defining `GrpcEnableMetricsIsSafe` in the header, then calling it here.

File: google/cloud/storage/internal/grpc/metrics_exporter_impl.cc [hunk 4]
```diff
@@ -24,9 +24,21 @@
 #include "google/cloud/storage/options.h"
 #include "google/cloud/common_options.h"
 #include "google/cloud/grpc_options.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/log.h"
+#include "absl/strings/match.h"
+#include "absl/strings/string_view.h"
+#include "absl/strings/strip.h"
+#include <grpcpp/grpcpp.h>
+#if GOOGLE_CLOUD_CPP_STORAGE_AUTO_OTEL_METRICS
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/metrics_exporter_impl.cc:33]: redundant

File: google/cloud/storage/internal/grpc/metrics_exporter_impl.cc [hunk 5]
```diff
@@ -44,6 +56,29 @@ auto MakeReaderOptions(Options const& options) {
   return reader_options;
 }
 
+class ExporterRegistry {
+ public:
+  ExporterRegistry() = default;
+
+  static ExporterRegistry& Singleton() {
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/metrics_exporter_impl.cc:63]: to quote jgm: "Singleton..." \*shudders\*

File: google/cloud/storage/internal/grpc/metrics_exporter_impl.cc [hunk 6]
```diff
@@ -44,6 +56,29 @@ auto MakeReaderOptions(Options const& options) {
...
+    static auto* exporters = new ExporterRegistry;
+    return *exporters;
+  }
+
+  bool Register(std::string authority) {
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/metrics_exporter_impl.cc:68]: maybe a comment:

```cc
// Returns true if a new authority is registered, false if the authority has already been registered.
bool Register(...);
```

File: google/cloud/storage/internal/grpc/metrics_exporter_impl.cc [hunk 7]
```diff
@@ -63,9 +98,86 @@ absl::optional<ExporterConfig> MakeMeterProviderConfig(
...
+  auto exporter = otel_internal::MakeMonitoringExporter(
+      std::move(config.project),
+      monitoring_v3::MakeMetricServiceConnection(
+          std::move(config.exporter_connection_options)),
+      config.exporter_options);
+
+  auto provider = MakeGrpcMeterProvider(std::move(exporter),
+                                        std::move(config.reader_options));
+
+  auto const metrics = std::vector<absl::string_view>{
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/metrics_exporter_impl.cc:138]: err, this list will want to grow with gRPC. Is there a way to say "gimme all the metrics"? (assuming we want that)

File: google/cloud/storage/internal/storage_stub_factory.cc [hunk 8]
```diff
@@ -49,6 +49,10 @@ std::shared_ptr<grpc::Channel> CreateGrpcChannel(
   // to send and/or receive those queries.
   args.SetInt(GRPC_ARG_DNS_ENABLE_SRV_QUERIES, 0);
 
+  if (options.has<AuthorityOption>()) {
```
Comment (dbolduc) [inline google/cloud/storage/internal/storage_stub_factory.cc:52]: unrelated. send a separate PR?

File: google/cloud/storage/internal/grpc/default_options_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc/default_options_test.cc]: Done

File: google/cloud/storage/internal/grpc/metrics_exporter_impl.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc/metrics_exporter_impl.cc]: Removed.

File: google/cloud/storage/internal/grpc/metrics_exporter_impl.cc (hunk 6 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc/metrics_exporter_impl.cc:68]: Done.

File: google/cloud/storage/internal/grpc/metrics_exporter_impl.cc (hunk 7 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc/metrics_exporter_impl.cc:135]: There is no way to say "all metrics" that I could find.  Some of these are not even listed in any header files, you just have to know they exist.

File: google/cloud/storage/internal/grpc/metrics_exporter_impl.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc/metrics_exporter_impl.cc:61]: This should go away in gRPC 1.65 or 1.66. I expect we will be able to configure these per-channel.

File: google/cloud/storage/internal/storage_stub_factory.cc (hunk 8 above)
Comment (coryan) [reply google/cloud/storage/internal/storage_stub_factory.cc:53]: It is related, the code in metrics_exporter_impl does not work without this change, and that is the only motivation for this change.

File: google/cloud/storage/internal/grpc/default_options.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc/default_options.cc]: My bad, fixed.

File: google/cloud/storage/google_cloud_cpp_storage_grpc.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/storage/google_cloud_cpp_storage_grpc.cmake:248]: up to you.

**Review by dbolduc — APPROVED** [review]
